Documentation/templates/ib/segues-swift3.md

```diff
@@ -0,0 +1,58 @@
+## Template Information
+
+| Name      | Description       |
+| --------- | ----------------- |
+| File name | ib/segues-swift3.stencil |
+| Invocation example | `swiftgen ib -t segues-swift3 …` |
+| Language | Swift 3 |
+| Author | Olivier Halligon |
+
+## When to use it
+
+- When you need to generate *Swift 3* code for your storyboard *segues*.
+- The generated code supports both UIKit platforms (iOS, tvOS and watchOS) and AppKit platform (macOS).
+
+## Customization
+
+You can customize some elements of this template by overriding the following parameters when invoking `swiftgen` in the command line, using `--param <paramName>=<newValue>`
+
+| Parameter Name | Default Value | Description |
+| -------------- | ------------- | ----------- |
+| `enumName` | `StoryboardSegue` | Allows you to change the name of the generated `enum` containing all storyboard segues. |
+| `module` | N/A | By default, the template will import the needed modules for custom classes, but won’t import the target’s module to avoid an import warning — using the `PRODUCT_MODULE_NAME` environment variable to detect it. Should you need to ignore an additional module, you can provide it here. |
+| `ignoreTargetModule` | N/A | Setting this parameter will disable the behaviour of prefixing classes with their module name for (only) the target module. |
+| `publicAccess` | N/A | If set, the generated constants will be marked as `public`. Otherwise, they'll be declared `internal`. |
+
+## Generated Code
+
+Note: the generated code may look differently depending on the platform the storyboard file is targeting.
+
+**Extract:**
+
+```swift
+enum StoryboardSegue {
+  enum Message: String, SegueType {
+    case embed
+    case nonCustom
+  }
+}
+```
+
+[Full generated code](https://github.com/SwiftGen/templates/blob/master/Tests/Fixtures/Generated/IB-iOS/swift3-context-all.swift)
+
+## Usage example
+
+```swift
+// You can perform segues using:
+vc.perform(segue: StoryboardSegue.Message.embed)
+
+// or match them (in prepareForSegue):
+override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+  switch StoryboardSegue.Message(rawValue: segue.identifier!)! {
```

Comment:
I know there's your PR on improved segues pending, but wondering if it would be complicated to just add:

```swift
extension SegueType {
  init(_ segue: UIStoryboardSegue) {
    guard let identifier = segue.identifier else { fatalError("Segue has no identifier") }
    guard let value = self.init(rawValue: identifier) else { fatalError("Segue with identifier \(identifier) wasn't of the expected type \(Self)") }
  self = value
  }
}
```
So that at least the call site / usage example would have less force-unwraps, like this:
```
override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
  switch StoryboardSegue.Message(segue) {
  case .embed: …
  case .nonCustom: …
  }
}
```
After all, SwiftGen is all about avoiding string-based identifiers… but also about avoiding force-casts and force-unwraps at call site, so… :wink:

While waiting for your improved Segues PRs that would at least easily improve the situation in the meantime, without too much code to add in the current templates?

Reply:
Oooh, I would not do that, that's a bad idea. This would break SwiftGen segues for a lot of people. You can't assume that all your segues have names, that's exactly why I have the extra `unnamedSegue` in the other PR.

I know that this would break quite a few of the apps we have here, because sometimes we name all the segues, but sometimes we just try to cast the destination to a type and set properties (for example). In those cases, segues don't always have names. I'm not discussing if all segues **must** have names, just saying that they don't always have a name. 

What we could do, but TBH that would belong in a separate PR (because this is just the split, no template modifications), is cherry-pick the `unnamedSegue` improvement from the other PR, and then we can do your idea (without throwing `fatalError`s for a whole bunch of users).

Reply:
Yes actually, this is what I currently use:
```swift
switch StoryboardSegue.ProductDetails(rawValue: segue.identifier ?? "") {
case .infoHeader?:
	... stuff ...
default:
	break
}
```

If we add the `unnamedSegue` case, and your first `init` (lightly modified), it becomes:
```swift
extension SegueType {
  init(_ segue: UIStoryboardSegue) {
    guard let identifier = segue.identifier else { return .unnamedSegue }
    guard let value = self.init(rawValue: identifier) else { fatalError("Segue with identifier \(identifier) wasn't of the expected type \(Self)") }
    return value
  }
}

switch StoryboardSegue.ProductDetails(segue) {
case .infoHeader:
	... stuff ...
case .unnamedSegue: // or default
	break
}
```

Reply:
We then should only generate the `unnamedSegue` case only of there exist such a segue in the storyboard. If all segues have an identifier then that extra case wouldn't make sense.

Maybe that's already what you do in your other PR idk. In the meantime, returning `nil` instead of `.unnamedSegue` for now (maybe until we re-discuss this in your PR for specific cases?) in the guard would feel more logical and simple for this PR until we can be sure, probably via your PR i the future, that unnamedSegue is only generated if such an unnamed segue exists.

Reply:
Yeah #364 makes the necessary modifications so that unnamed segues are also in the context instead of being filtered out. The templates are also modified to take this into account. In that sense #364 is a breaking change BTW.

To clarify, given what this PR has now, the only thing we can provide is the following `init`:

Reply:
Cool. Let's do that then 😉
(Can't wait for #364 btw)

Reply:
Yeah, I discovered yesterday that there are ...issues... when we combine #364 with #485, specifically, the parser can't build up the correct `custom types` set because some classes/modules aren't known at parse time. But that discussion should go in the other PR.

@AliSoftware Quick question though: should I cherry-pick the "don't filter out unnamed segues" out of #364 into a separate PR, that can be merged in 6.0, so we don't have a breaking change for #485?

Reply:
Yeah, probably, especially if that `init` will then "support" them by returning `nil`.

As of today, people do use them as such anyway: if they know they have unnamed segues, they'll `guard let id = sender.identifier` manually, and if they don't (or are careless) they'll `StoryboardSegue.Message(sender.identifier!)`, but that's their choice of crashing and not accounting for unnamed segues then, and with not filtering the unnamed segues anymore that `init` will at least give them a nicer solution to avoid a crash.